cmd/sync-rover-groups/yamlgroupcollector.go

```diff
@@ -0,0 +1,206 @@
+						if err != nil {
+							logrus.WithField("source-file", path).WithError(err).Error("Failed to get groups")
+							return err
+						}
+						groups = groups.Union(more)
+					}
+				case *templatev1.Template:
+					for _, object := range o.Objects {
+						subObj, _, err := c.decoder.Decode(object.Raw, nil, nil)
+						if err != nil {
+							if runtime.IsNotRegisteredError(err) && (!c.validateSubjects ||
+								!strings.HasPrefix(err.Error(), "no kind \"Group\" is registered for version \"v1\"")) {
+								continue
+							}
+							// Trying to avoid reinventing oc-process
+							// We might miss some groups but users can make RBACs out of templates
+							// Most commonly, it is because of `replicas: ${{REPLICAS}}`
+							logrus.WithField("source-file", path).WithError(err).Warning("Failed to decode sub-object in template")
+							continue
+						}
+						more, err := primitive(subObj, c.validateSubjects)
+						if err != nil {
+							logrus.WithField("source-file", path).WithError(err).Error("Failed to get groups")
+							return err
+						}
+						groups = groups.Union(more)
+					}
+				default:
+					more, err := primitive(obj, c.validateSubjects)
+					if err != nil {
+						logrus.WithField("source-file", path).WithError(err).Error("Failed to get groups")
+						return err
+					}
+					groups = groups.Union(more)
+				}
+			}
+
+			return nil
+		}); err != nil {
+		return nil, fmt.Errorf("failed to walk dir: %w", err)
+
+	}
+	return groups, nil
+}
+
+func onlyComments(yaml []byte) bool {
+	for _, line := range bytes.Split(yaml, []byte("\n")) {
+		if bytes.Equal(bytes.TrimSpace(line), []byte("")) {
+			continue
+		}
+		if !bytes.HasPrefix(line, []byte("#")) {
+			return false
+		}
+	}
+	return true
+}
+
+func primitive(obj runtime.Object, validateSubjects bool) (sets.String, error) {
+	groups := sets.NewString()
+	switch o := obj.(type) {
+	case *userv1.Group:
+		if validateSubjects {
+			return nil, fmt.Errorf("cannot create any group but found: %s", o.Name)
+		}
+	case *rbacv1.RoleBinding:
+		for _, s := range o.Subjects {
+			if validateSubjects && s.Kind == "User" {
+				return nil, fmt.Errorf("cannot use User as subject in RoleBinding: %s", o.Name)
+			}
+			if s.Kind == "Group" {
+				if strings.HasPrefix(s.Name, "system:") {
+					continue
+				}
+				if strings.Contains(s.Name, "${") {
+					return nil, fmt.Errorf("cannot use ${ in a subject of RoleBinding: %s", o.Name)
+				}
+				groups.Insert(s.Name)
+			}
+		}
+	case *rbacv1.ClusterRoleBinding:
+		for _, s := range o.Subjects {
+			if validateSubjects && s.Kind == "User" {
+				return nil, fmt.Errorf("cannot use User as subject in ClusterRoleBinding: %s", o.Name)
+			}
+			if s.Kind == "Group" {
+				if strings.HasPrefix(s.Name, "system:") {
+					continue
+				}
+				if strings.Contains(s.Name, "${") {
+					return nil, fmt.Errorf("cannot use ${ in a subject of ClusterRoleBinding: %s", o.Name)
+				}
+				groups.Insert(s.Name)
+			}
```

Comment:
Can these two blocks share code? Looks like the only difference is the `{Cluster}RoleBinding` bit in the errors...

Reply:
Updated.